In [949]:
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np
import csv
import plotly
import plotly.graph_objs as go
import plotly.express as px
from plotly.subplots import make_subplots
from sklearn.metrics import r2_score

In [950]:
data = pd.read_csv('Кинетика.csv', delimiter=';')
data

t.min    [A]    [B]   [A]exp   [B]exp  t.min.1  [A].1  [B].1  [A]exp.1  \
0    0.000  0.003  0.160  0.00297  0.16160    0.000  0.004  0.160   0.00404   
1    0.001  0.003  0.160  0.00303  0.15840    0.001  0.004  0.160   0.00400   
2    0.002  0.003  0.160  0.00303  0.16160    0.002  0.004  0.160   0.00404   
3    0.003  0.003  0.160  0.00300  0.15840    0.003  0.004  0.160   0.00396   
4    0.004  0.003  0.160  0.00300  0.15840    0.004  0.004  0.160   0.00396   
..     ...    ...    ...      ...      ...      ...    ...    ...       ...   
995  0.995  0.000  0.157  0.00000  0.15543    0.995  0.000  0.156   0.00000   
996  0.996  0.000  0.157  0.00000  0.15543    0.996  0.000  0.156   0.00000   
997  0.997  0.000  0.157  0.00000  0.15700    0.997  0.000  0.156   0.00000   
998  0.998  0.000  0.157  0.00000  0.15857    0.998  0.000  0.156   0.00000   
999  0.999  0.000  0.157  0.00000  0.15543    0.999  0.000  0.156   0.00000   

     [B]exp.1  ...  t.min.8  [A].8  [B].8  [A]exp.8  [B]exp.8  t.min.9  [A].9  \
0     0.16000  ...    0.000  0.160  0.006   0.16160   0.00606    0.000  0.160   
1     0.15840  ...    0.001  0.160  0.006   0.15840   0.00606    0.001  0.160   
2     0.16160  ...    0.002  0.160  0.006   0.16160   0.00594    0.002  0.160   
3     0.16000  ...    0.003  0.160  0.006   0.15840   0.00606    0.003  0.160   
4     0.15840  ...    0.004  0.160  0.006   0.16160   0.00606    0.004  0.160   
..        ...  ...      ...    ...    ...       ...       ...      ...    ...   
995   0.15444  ...    0.995  0.154  0.000   0.15400   0.00000    0.995  0.153   
996   0.15756  ...    0.996  0.154  0.000   0.15554   0.00000    0.996  0.153   
997   0.15444  ...    0.997  0.154  0.000   0.15554   0.00000    0.997  0.153   
998   0.15600  ...    0.998  0.154  0.000   0.15246   0.00000    0.998  0.153   
999   0.15444  ...    0.999  0.154  0.000   0.15554   0.00000    0.999  0.153   

     [B].9  [A]exp.9  [B]exp.9  
0    0.007   0.15840   0.00693  
1    0.007   0.15840   0.00707  
2    0.007   0.16160   0.00700  
3    0.007   0.16160   0.00707  
4    0.007   0.15840   0.00707  
..     ...       ...       ...  
995  0.000   0.15147   0.00000  
996  0.000   0.15147   0.00000  
997  0.000   0.15300   0.00000  
998  0.000   0.15300   0.00000  
999  0.000   0.15300   0.00000  

[1000 rows x 50 columns]

In [951]:
x = data['t.min'].head(180)
y = data['[A]exp'].head(180)
fig = px.scatter(data_frame=data, x=x, y=y, title='[A](t) для вычисления начальной скорости 1')

#Построение линии тренда с фиксированным свободным членом равным начальной концентрации
coeffs = np.polyfit(x, y, deg=1)
coeffs[1] = data['[A]exp'].values[0]
poly_eq = np.poly1d(coeffs)
x_trend = np.linspace(x.min(), x.max(), 100)
y_trend = poly_eq(x_trend)
fig.add_scatter(x=x_trend, y=y_trend, mode='lines', name='Линия тренда', line=dict(color='red'))

y_pred = poly_eq(x) 
r_squared = r2_score(y, y_pred)


print('Коэффициенты уравнения линии тренда: a = ' + str(coeffs[0])+', b = '+str(coeffs[1]))
print('R^2: '+str(r_squared))

fig.show()
r1 = -(coeffs[0])
print('r1 = '+str(r1))

Коэффициенты уравнения линии тренда: a = -0.007430105867465048, b = 0.00297
R^2: 0.6606652188729265


r1 = 0.007430105867465048


In [952]:
x = data['t.min.1'].head(180)
y = data['[A]exp.1'].head(180)
fig = px.scatter(data_frame=data, x=x, y=y, title='[A](t) для вычисления начальной скорости 2')

#Построение линии тренда с фиксированным свободным членом равным начальной концентрации
coeffs = np.polyfit(x, y, deg=1)
coeffs[1] = data['[A]exp.1'].values[0]
poly_eq = np.poly1d(coeffs)
x_trend = np.linspace(x.min(), x.max(), 100)
y_trend = poly_eq(x_trend)
fig.add_scatter(x=x_trend, y=y_trend, mode='lines', name='Линия тренда', line=dict(color='red'))

y_pred = poly_eq(x) 
r_squared = r2_score(y, y_pred)


print('Коэффициенты уравнения линии тренда: a = ' + str(coeffs[0])+', b = '+str(coeffs[1]))
print('R^2: '+str(r_squared))

fig.show()
r2 = -(coeffs[0])
print('r2 = '+str(r2))

Коэффициенты уравнения линии тренда: a = -0.01010429334238712, b = 0.00404
R^2: 0.741266433371618


r2 = 0.01010429334238712


In [953]:
x = data['t.min.2'].head(180)
y = data['[A]exp.2'].head(180)
fig = px.scatter(data_frame=data, x=x, y=y, title='[A](t) для вычисления начальной скорости 3')

#Построение линии тренда с фиксированным свободным членом равным начальной концентрации
coeffs = np.polyfit(x, y, deg=1)
coeffs[1] = data['[A]exp.2'].values[0]
poly_eq = np.poly1d(coeffs)
x_trend = np.linspace(x.min(), x.max(), 100)
y_trend = poly_eq(x_trend)
fig.add_scatter(x=x_trend, y=y_trend, mode='lines', name='Линия тренда', line=dict(color='red'))

y_pred = poly_eq(x) 
r_squared = r2_score(y, y_pred)


print('Коэффициенты уравнения линии тренда: a = ' + str(coeffs[0])+', b = '+str(coeffs[1]))
print('R^2: '+str(r_squared))

fig.show()
r3 = -(coeffs[0])
print('r3 = '+str(r3))

Коэффициенты уравнения линии тренда: a = -0.012840828420630283, b = 0.00505
R^2: 0.8474643579061865


r3 = 0.012840828420630283


In [954]:
x = data['t.min.3'].head(180)
y = data['[A]exp.3'].head(180)
fig = px.scatter(data_frame=data, x=x, y=y, title='[A](t) для вычисления начальной скорости 4')

#Построение линии тренда с фиксированным свободным членом равным начальной концентрации
coeffs = np.polyfit(x, y, deg=1)
coeffs[1] = data['[A]exp.3'].values[0]
poly_eq = np.poly1d(coeffs)
x_trend = np.linspace(x.min(), x.max(), 100)
y_trend = poly_eq(x_trend)
fig.add_scatter(x=x_trend, y=y_trend, mode='lines', name='Линия тренда', line=dict(color='red'))

y_pred = poly_eq(x) 
r_squared = r2_score(y, y_pred)


print('Коэффициенты уравнения линии тренда: a = ' + str(coeffs[0])+', b = '+str(coeffs[1]))
print('R^2: '+str(r_squared))

fig.show()
r4 = -(coeffs[0])
print('r4 = '+str(r4))

Коэффициенты уравнения линии тренда: a = -0.01302423943125818, b = 0.006
R^2: 0.7871621465683692


r4 = 0.01302423943125818


In [955]:
x = data['t.min.4'].head(180)
y = data['[A]exp.4'].head(180)
fig = px.scatter(data_frame=data, x=x, y=y, title='[A](t) для вычисления начальной скорости 5')

#Построение линии тренда с фиксированным свободным членом равным начальной концентрации
coeffs = np.polyfit(x, y, deg=1)
coeffs[1] = data['[A]exp.4'].values[0]
poly_eq = np.poly1d(coeffs)
x_trend = np.linspace(x.min(), x.max(), 100)
y_trend = poly_eq(x_trend)
fig.add_scatter(x=x_trend, y=y_trend, mode='lines', name='Линия тренда', line=dict(color='red'))

y_pred = poly_eq(x) 
r_squared = r2_score(y, y_pred)


print('Коэффициенты уравнения линии тренда: a = ' + str(coeffs[0])+', b = '+str(coeffs[1]))
print('R^2: '+str(r_squared))

fig.show()
r5 = -(coeffs[0])
print('r5 = '+str(r5))

Коэффициенты уравнения линии тренда: a = -0.01721721863843535, b = 0.007
R^2: 0.89831429990136


r5 = 0.01721721863843535


In [956]:
x = data['t.min.5'].head(180)
y = data['[B]exp.5'].head(180)
fig = px.scatter(data_frame=data, x=x, y=y, title='[B](t) 6')

#Построение линии тренда с фиксированным свободным членом равным начальной концентрации
coeffs = np.polyfit(x, y, deg=1)
coeffs[1] = data['[B]exp.5'].values[0]
poly_eq = np.poly1d(coeffs)
x_trend = np.linspace(x.min(), x.max(), 100)
y_trend = poly_eq(x_trend)
fig.add_scatter(x=x_trend, y=y_trend, mode='lines', name='Линия тренда', line=dict(color='red'))

y_pred = poly_eq(x) 
r_squared = r2_score(y, y_pred)


print('Коэффициенты уравнения линии тренда: a = ' + str(coeffs[0])+', b = '+str(coeffs[1]))
print('R^2: '+str(r_squared))

fig.show()
r6 = -(coeffs[0])
print('r6 = '+str(r6))

Коэффициенты уравнения линии тренда: a = -0.007348066298342542, b = 0.00297
R^2: 0.6605579533907506


r6 = 0.007348066298342542


In [957]:
x = data['t.min.6'].head(180)
y = data['[B]exp.6'].head(180)
fig = px.scatter(data_frame=data, x=x, y=y, title='[B](t) 7')

#Построение линии тренда с фиксированным свободным членом равным начальной концентрации
coeffs = np.polyfit(x, y, deg=1)
coeffs[1] = data['[B]exp.6'].values[0]
poly_eq = np.poly1d(coeffs)
x_trend = np.linspace(x.min(), x.max(), 100)
y_trend = poly_eq(x_trend)
fig.add_scatter(x=x_trend, y=y_trend, mode='lines', name='Линия тренда', line=dict(color='red'))

y_pred = poly_eq(x) 
r_squared = r2_score(y, y_pred)


print('Коэффициенты уравнения линии тренда: a = ' + str(coeffs[0])+', b = '+str(coeffs[1]))
print('R^2: '+str(r_squared))

fig.show()
r7 = -(coeffs[0])
print('r7 = '+str(r7))

Коэффициенты уравнения линии тренда: a = -0.01015852341121641, b = 0.00404
R^2: 0.7437946464344931


r7 = 0.01015852341121641


In [958]:
x = data['t.min.7'].head(180)
y = data['[B]exp.7'].head(180)
fig = px.scatter(data_frame=data, x=x, y=y, title='[B](t) 8')

#Построение линии тренда с фиксированным свободным членом равным начальной концентрации
coeffs = np.polyfit(x, y, deg=1)
coeffs[1] = data['[B]exp.7'].values[0]
poly_eq = np.poly1d(coeffs)
x_trend = np.linspace(x.min(), x.max(), 100)
y_trend = poly_eq(x_trend)
fig.add_scatter(x=x_trend, y=y_trend, mode='lines', name='Линия тренда', line=dict(color='red'))

y_pred = poly_eq(x) 
r_squared = r2_score(y, y_pred)


print('Коэффициенты уравнения линии тренда: a = ' + str(coeffs[0])+', b = '+str(coeffs[1]))
print('R^2: '+str(r_squared))

fig.show()
r8 = -(coeffs[0])
print('r8 = '+str(r8))

Коэффициенты уравнения линии тренда: a = -0.012715958311470524, b = 0.00505
R^2: 0.8403258448423244


r8 = 0.012715958311470524


In [959]:
x = data['t.min.8'].head(180)
y = data['[B]exp.8'].head(180)
fig = px.scatter(data_frame=data, x=x, y=y, title='[B](t) 9')

#Построение линии тренда с фиксированным свободным членом равным начальной концентрации
coeffs = np.polyfit(x, y, deg=1)
coeffs[1] = data['[B]exp.8'].values[0]
poly_eq = np.poly1d(coeffs)
x_trend = np.linspace(x.min(), x.max(), 100)
y_trend = poly_eq(x_trend)
fig.add_scatter(x=x_trend, y=y_trend, mode='lines', name='Линия тренда', line=dict(color='red'))

y_pred = poly_eq(x) 
r_squared = r2_score(y, y_pred)


print('Коэффициенты уравнения линии тренда: a = ' + str(coeffs[0])+', b = '+str(coeffs[1]))
print('R^2: '+str(r_squared))

fig.show()
r9 = -(coeffs[0])
print('r9 = '+str(r9))

Коэффициенты уравнения линии тренда: a = -0.013122442050680597, b = 0.00606
R^2: 0.7552901452242715


r9 = 0.013122442050680597


In [960]:
x = data['t.min.9'].head(180)
y = data['[B]exp.9'].head(180)
fig = px.scatter(data_frame=data, x=x, y=y, title='[B](t) 10')

#Построение линии тренда с фиксированным свободным членом равным начальной концентрации
coeffs = np.polyfit(x, y, deg=1)
coeffs[1] = data['[B]exp.9'].values[0]
poly_eq = np.poly1d(coeffs)
x_trend = np.linspace(x.min(), x.max(), 100)
y_trend = poly_eq(x_trend)
fig.add_scatter(x=x_trend, y=y_trend, mode='lines', name='Линия тренда', line=dict(color='red'))

y_pred = poly_eq(x) 
r_squared = r2_score(y, y_pred)


print('Коэффициенты уравнения линии тренда: a = ' + str(coeffs[0])+', b = '+str(coeffs[1]))
print('R^2: '+str(r_squared))

fig.show()
r10 = -(coeffs[0])
print('r10 = '+str(r10))

Коэффициенты уравнения линии тренда: a = -0.01719026307396321, b = 0.00693
R^2: 0.8990564167800594


r10 = 0.01719026307396321


In [961]:
x = [data['[A]exp'].values[0],data['[A]exp.1'].values[0], data['[A]exp.2'].values[0], data['[A]exp.3'].values[0], data['[A]exp.4'].values[0]]
y = [r1, r2, r3, r4, r5]
x_log = [np.log(i) for i in x]
y_log = [np.log(i) for i in y]
print(x)
print(y)
print(x_log)
print(y_log)
fig = px.scatter(data_frame=data, x=x_log, y=y_log, title='lnr(ln[A]0)')
fig.update_layout(xaxis_title='ln[A]0', yaxis_title='ln(r)')

coeffs = np.polyfit(x_log, y_log, deg=1)
poly_eq = np.poly1d(coeffs)
x_trend = np.linspace(x_log[0], x_log[4], 100)
y_trend = poly_eq(x_trend)
fig.add_scatter(x=x_trend, y=y_trend, mode='lines', name='Линия тренда', line=dict(color='red'))

y_pred = poly_eq(x_log) 
r_squared = r2_score(y_log, y_pred)



print('Коэффициенты уравнения линии тренда: a = ' + str(coeffs[0])+', b = '+str(coeffs[1]))
print('R^2: '+str(r_squared))

fig.show()


[0.00297, 0.00404, 0.00505, 0.006, 0.007]
[0.007430105867465048, 0.01010429334238712, 0.012840828420630283, 0.01302423943125818, 0.01721721863843535]
[-5.819193326167529, -5.511510587009078, -5.288367035694868, -5.115995809754082, -4.961845129926823]
[-4.902215171704579, -4.594794862052613, -4.35512546406089, -4.340943086041258, -4.061845312282817]
Коэффициенты уравнения линии тренда: a = 0.9181346478838615, b = 0.45128718004806917
R^2: 0.9658382988975578


In [962]:
print('n = '+str(coeffs[0]))
print('k = '+str(np.exp(coeffs[1])))
n0 = (np.log(r2/r1))/np.log(data['[A]exp.1'].values[0]/data['[A]exp'].values[0])
print('порядок по методу Вант-Гоффа = '+str(n0))

n = 0.9181346478838615
k = 1.570332185417377
порядок по методу Вант-Гоффа = 0.999147077579972


In [963]:
x = [data['[B]exp.5'].values[0],data['[B]exp.6'].values[0], data['[B]exp.7'].values[0], data['[B]exp.8'].values[0], data['[B]exp.9'].values[0]]
y = [r6, r7, r8, r9, r10]
x_log = [np.log(i) for i in x]
y_log = [np.log(i) for i in y]
print(x)
print(y)
print(x_log)
print(y_log)
fig = px.scatter(data_frame=data, x=x_log, y=y_log, title='lnr(ln[B]0)')
fig.update_layout(xaxis_title='ln[B]0', yaxis_title='ln(r)')

coeffs = np.polyfit(x_log, y_log, deg=1)
poly_eq = np.poly1d(coeffs)
x_trend = np.linspace(x_log[0], x_log[4], 100)
y_trend = poly_eq(x_trend)
fig.add_scatter(x=x_trend, y=y_trend, mode='lines', name='Линия тренда', line=dict(color='red'))

y_pred = poly_eq(x_log) 
r_squared = r2_score(y_log, y_pred)



print('Коэффициенты уравнения линии тренда: a = ' + str(coeffs[0])+', b = '+str(coeffs[1]))
print('R^2: '+str(r_squared))

fig.show()

[0.00297, 0.00404, 0.00505, 0.00606, 0.00693]
[0.007348066298342542, 0.01015852341121641, 0.012715958311470524, 0.013122442050680597, 0.01719026307396321]
[-5.819193326167529, -5.511510587009078, -5.288367035694868, -5.1060454789009135, -4.971895465780325]
[-4.913318089032163, -4.5894421809354355, -4.364897514370024, -4.333431380874958, -4.063412155809008]
Коэффициенты уравнения линии тренда: a = 0.9291325641785465, b = 0.5081123591080393
R^2: 0.9649044086373213


In [964]:
print('n = '+str(coeffs[0]))
print('k = '+str(np.exp(coeffs[1])))

n = 0.9291325641785465
k = 1.6621506881962833


Суммарный порядок реакции равен 2 т.к. n = 0.92 ~ 1 для двух расходующихся реагентов


Константа скорости реакции равна ~ 1.6